docs/stdenv: Document updateScript features (#200951)
NixOS/nixpkgs: doc/stdenv/stdenv.chapter.md

    update-source-version zoom-us "$version"
  '';
  ```

- a list, a script followed by arguments to be passed to it:

  ```nix
  passthru.updateScript = [ ../../update.sh pname "--requested-release=unstable" ];
  ```

- an attribute set containing:
  - [`command`]{#var-passthru-updateScript-set-command} – a string or list in the [format expected by `passthru.updateScript`](#var-passthru-updateScript-command).
  - [`attrPath`]{#var-passthru-updateScript-set-attrPath} (optional) – a string containing the canonical attribute path for the package. If present, it will be passed to the update script instead of the attribute path on which the package was discovered during Nixpkgs traversal.
  - [`supportedFeatures`]{#var-passthru-updateScript-set-supportedFeatures} (optional) – a list of the [extra features](#var-passthru-updateScript-supported-features) the script supports.

  ```nix
  passthru.updateScript = {
    command = [ ../../update.sh pname ];
    attrPath = pname;
    supportedFeatures = [ … ];
  };
  ```

##### How update scripts are executed? {#var-passthru-updateScript-execution}

> Update scripts are to be invoked by `maintainers/scripts/update.nix` script. You can run `nix-shell maintainers/scripts/update.nix` in the root of Nixpkgs repository for information on how to use it. `update.nix` offers several modes for selecting packages to update (e.g. select by attribute path, traverse Nixpkgs and filter by maintainer…), and it will execute update scripts for all matched packages that have an `updateScript` attribute.

Reviewer: ```suggestion
Update scripts are to be invoked by `maintainers/scripts/update.nix` script.
You can run `nix-shell maintainers/scripts/update.nix` in the root of Nixpkgs repository for information on how to use it.
`update.nix` offers several modes for selecting packages to update (e.g. select by attribute path, traverse Nixpkgs and filter by maintainer, etc.), and it will execute update scripts for all matched packages that have an `updateScript` attribute.
```

Author: Thanks, replaced ellipsis with etc.

passthru.updateScript = [ ../../update.sh pname "--requested-release=unstable" ];
```

> Finally, the attribute can be an attribute set, which, in addition to the command, can optionally provide the canonical attribute path to use for the list of extra features the script supports.

Reviewer: Don't use "finally" or other indication of ordering in prose (same as we set `1.` for all numbered list elements in markdown) - order may change, and then the prose will have to change.

```suggestion
The attribute can be an attribute set. In addition to the command, it can optionally provide the canonical attribute path to use for the list of extra features the script supports.
```

Other than that, I don't understand that sentence. As I read it right now, `pname` is supposed to be somehow used "in the list of extra features". But how? The example doesn't show. Why? This is not explained here or in the next section.

I'd suggest to rather list the attributes and note what exactly each one is for, and then make a full example that actually works. It would be OK to reference an existing update script (link to source so it can be easily found for inspection) or referring to an example from previous sections.

```nix
passthru.updateScript = {
  command = [ ../../update.sh pname ];
  attrPath = pname;
  supportedFeatures = [ … ];
};
```

##### How update scripts are executed? {#var-passthru-updateScript-execution}

> For information about how to run update scripts, run `nix-shell maintainers/scripts/update.nix` in the root of Nixpkgs repository. The script offers several modes for finding and executing update scripts.

Reviewer: Which are these modes? Also please link to `update.nix`, otherwise there are additional steps to find out what it actually does, since it's not documented here.

> Each script will be run with the `UPDATE_NIX_NAME`, `UPDATE_NIX_PNAME`, `UPDATE_NIX_OLD_VERSION` and `UPDATE_NIX_ATTR_PATH` environment variables set respectively to the name, pname, old version and attribute path of the package it is supposed to update.

Reviewer: This would be much easier to read as a list with annotations. Please fill in something reasonable, I just made things up to show the idea.

```suggestion
Each script will be run with the following environment variables set according to the package being updated:

 - `UPDATE_NIX_NAME` - ???
 - `UPDATE_NIX_PNAME` - package name
 
   Example: `clang`
   
-  `UPDATE_NIX_OLD_VERSION` - old package version to update

   Example: `11.1.0`
   
- `UPDATE_NIX_ATTR_PATH` - attribute path of the package

   Example: `llvmPackages.clang-unwrapped`
   
```

The term "attribute path" apparently is nowhere really defined except implicitly in [`attrsets.nix`](https://github.com/NixOS/nixpkgs/blob/5bc1b01a403e071500a3b0659364fb6ee20391e1/lib/attrsets.nix#L15-L30), one currently has to know that it means roughly "a sequence of attribute names to resolve to get to that value". In this context it's also missing that the attribute this path belongs to, which is probably the one in `all-packages.nix`, but how would one know exactly?

::: {.note}
The script will be usually run from the root of the Nixpkgs repository but you should not rely on that. Also note that the update scripts will be run in parallel by default; you should avoid running `git commit` or any other commands that cannot handle that.
:::

`maintainers/scripts/update.nix` also supports automatically creating commits when running it with `--argstr commit true`. Update scripts can opt to override the commit contents by adding `commit` to `supportedFeatures` and [modifying the script accordingly](#var-passthru-updateScript-commit). It might be useful, for example, for including news link in the commit message.

##### Supported features {#var-passthru-updateScript-supported-features}

> ###### `commit` {#var-passthru-updateScript-commit}

Author: Maybe we should rename the feature to something like `commitProtocol`, as it might be confusing (e.g. https://github.com/NixOS/nixpkgs/pull/193248#discussion_r981779706).

Reviewer 2: This explanation could be more verbose. I had difficulty in understanding the idea initially. Maybe the explanation found in the linked comment can be added to this PR.

Author: @romildo I have added a introductory sentence:

> This feature allows update scripts to *ask* `update.nix` to create Git commits.

And mentioned again in the tip that update scripts should not actually create commits.

When support of this feature is declared, whenever the update script exits with `0` return status, it is expected to print a JSON list containing an object for each updated attribute. When `update.nix` is run in commit mode, each of the objects will result in a separate commit being created. Empty list can be returned when the script did not update any files, for example, when the package is already at the latest version.

Each object must contain the keys shown below:

```json
[
  {
    "attrPath": "volume_key",
    "oldVersion": "0.3.11",
    "newVersion": "0.3.12",
    "files": [
      "/path/to/nixpkgs/pkgs/development/libraries/volume-key/default.nix"
    ]
  }
]
```

> Unless the returned array contains exactly one object (e.g. `[{}]`), in which case all keys can be omitted and will be determined automatically.

Reviewer: What does that mean?

> In addition to the keys listed above `commitBody` key can be provided, which will append extra content to the default commit message (useful for adding changelog links). Or `commitMessage` key can override the commit message completely.

Reviewer: As above, this should be a list of keys, each with a description, followed by a full example of how they are used. It's much harder to read prose than skim a list, and no one really reads things when looking up information.

Author: Done.
